game-core/src/main/java/games/strategy/triplea/attachments/UnitAttachment.java

```diff
@@ -1751,20 +1756,54 @@ private void resetCanDieFromReachingMaxDamage() {
     canDieFromReachingMaxDamage = false;
   }
 
+  @Deprecated
   private void setIsSuicide(final String s) {
-    isSuicide = getBool(s);
+    setIsSuicide(getBool(s));
   }
 
+  @Deprecated
   private void setIsSuicide(final Boolean s) {
-    isSuicide = s;
+    setIsSuicideOnAttack(s);
+    // Global property controlled whether isSuicide units would suicide on defense
+    setIsSuicideOnDefense(s && !Properties.getDefendingSuicideAndMunitionUnitsDoNotFire(getData()));
+    setIsFirstStrike(s);
+    if (s) {
+      canNotTarget = null;
```

Comment:
The set to null might be potentially dangerous / NPE if `setCanNotTarget` is invoked after `setIsSuicide`, noting on line 683:

```
  private void setCanNotTarget(final String value) throws GameParseException {
    final String[] s = splitOnColon(value);
    for (final String u : s) {
      final UnitType ut = getData().getUnitTypeList().getUnitType(u);
      if (ut == null) {
        throw new GameParseException("canNotTarget: no such unit type: " + u + thisErrorMsg());
      }
683:      canNotTarget.add(ut);
    }
  }
```

Would that be a potential problem?

Reply:
Second, the set to null seems to rely that any access to `canNotTarget` will go through a `get` method which has the side effect of also setting `canNotTarget`. Could this be fixed and/or be made more explicit by doing an assignment here of that value? (this seems kinda hacky, but it perhaps it makes things explicit at least):

```
if(s) {
  canNotTarget = getCanNotTarget();
}
```

Reply:
setCanNotTarget and setIsSuicide should never be used in the same map. setIsSuicide is just for existing map compatibility.

That's correct. The reason for this is you can't set the canNotTarget until all the XML unit attachments are parsed (as you need to read some of the unit properties) so has to happen essentially after all unit parsing is done. This was the easiest way I could think to do it and the only way to access it is through the getter (null is essentially being used as a place holder to flag that canNotTarget needs to still be calculated). This is a bit hacky but without adding some sort of placeholder state or extra parsing steps, I couldn't think of a better way to do it.

Reply:
I understand that `setCanNotTarget` and `setIsSuicide` are now being considered incompatible settings. Will the failure mode when that is done be a NPE? I suspect eventually, on some map, we'll see the two settings be combined and the NPE would be an unhelpful error message.

If we could get it so that the system fails with an exact error message (EG: unsupported option combination) and tells the user what/where, that would be quite valuable and we would handling this situation well enough.  

Could we possibly achieve that by adding a null-guard in the 'setCanNotTarget' method and then adding a post-processing check if the two settings have been used together? Second question, do we really need to set `canNotTarget` to null? If we can avoid the null, then it may simplify the need for null-checks and perhaps we could have the two options exist with each other, but perhaps the behavior would be non-deterministic.

Reply:
Yeah, you'll get a NPE pointing to that method which will make it pretty obvious what the issue is. I don't expect it to ever happen and it would only be a map maker which would see it and probably just post on the forum. We can add checks to try to check for the null and return a better message just wondering if we are over engineering something that I don't think will ever happen.

You could avoid setting the null by setting some other say boolean field that then is checked instead of checking for the null in the getter though you are then adding additional temporary fields into a class that already has a lot of members variables. Not sure if that is really better or worse but was the other option I considered.

Reply:
It's always surprising the data combinations/states that eventually wind up happening. Realistically we could maybe see it if a map were being retrofitted, or an existing map copied and then updated to be something else and the map maker missed an occurrence of the extra option.

Here's perhaps an idea:
- check for a null value of `canNotTarget` in the setter, rather than NPE, throw an illegal state at the beginning of the setter. That would provide a way to give a direct message to indicate that an incompatible setting option was used.

Reply:
Added null check and custom exception message.